Enforce in bootstrap that check must have stage at least 1

Kobzol opened the pull request:
This PR is another step towards https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Proposal.20to.20cleanup.20stages.20and.20steps.20after.20the.20redesign/with/523586917, this time dealing with `x check`.

It enforces the invariant that:
- We check std stage N with rustc stage N
- We check everything else stage N with rustc stage N - 1
 
It creates a single function that prepares a proper build compiler for checking something, and also adds snapshot tests for various common check steps. Some obsolete code was also removed.

The default check stage also becomes 1, for all profiles. I tested manually that `x check std` with `download-ci-rustc` still works and doesn't build rustc locally.

Closes: https://github.com/rust-lang/rust/pull/139170

r? @ghost

Kobzol committed:
Make default check stage be 1, and error out on checking with stage 0

Kobzol committed:
Remove `custom_stage` override from `check::Std` and make 1 be the default check stage for it

Kobzol committed:
Add change tracker entry

Zalathar committed:
Fixup check of rust-analyzer, codegen backends, compiletest and other tools

Kobzol committed:
Implement `CoverageDump` checking through the `tool_check_step` macro

Kobzol committed:
Unify selection of build compiler for checking in a single function

Kobzol commented:
CC @jieyouxu (one of many heaps of bootstrap fixup commits I have in store :laughing:)

rust-log-analyzer commented:
The job **`mingw-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15896362010/job/44828842042) [(plain)](https://github.com/rust-lang/rust/commit/aed5f232749efdc6f3202f551cee97607adad7d6/checks/44828842042/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 3.131 Building wheels for collected packages: reuse
#16 3.132   Building wheel for reuse (pyproject.toml): started
#16 3.344   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.345   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#16 3.346   Stored in directory: /tmp/pip-ephem-wheel-cache-nsa9f00o/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.348 Successfully built reuse
#16 3.349 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.759 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.760 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.9s
```

</details>

Kobzol commented:
Cross-compilation checks are now doing too much work, I need to take a look at how we can reduce that, because `x check --stage1` with cross-compilation is no longer a thing.

rust-log-analyzer commented:
The job **`mingw-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15904675257/job/44856304524) [(plain)](https://github.com/rust-lang/rust/commit/a36ee2459c68b6f455b1f65af313534407957954/checks/44856304524/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.873 Building wheels for collected packages: reuse
#16 2.874   Building wheel for reuse (pyproject.toml): started
#16 3.090   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.091   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#16 3.091   Stored in directory: /tmp/pip-ephem-wheel-cache-dy_04mdo/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.093 Successfully built reuse
#16 3.093 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.494 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.495 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #143254) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kobzol commented:
r? @jieyouxu

rust-log-analyzer commented:
The job **`mingw-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16003998869/job/45146071135) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45146071135) [(plain)](https://github.com/rust-lang/rust/commit/22fb548879c094159c52b24d5ea7f539b31145ec/checks/45146071135/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 3.096 Building wheels for collected packages: reuse
#16 3.098   Building wheel for reuse (pyproject.toml): started
#16 3.311   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.312   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#16 3.312   Stored in directory: /tmp/pip-ephem-wheel-cache-4cv2rae2/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.314 Successfully built reuse
#16 3.315 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.711 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.712 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.8s
```

</details>

jieyouxu commented on src/bootstrap/src/core/builder/tests.rs:
```diff
@@ -1294,6 +1294,38 @@ mod snapshot {
+        insta::assert_snapshot!(
+            ctx.config("check")
+                .stage(2)
+                .targets(&[TEST_TRIPLE_1])
+                .hosts(&[TEST_TRIPLE_1])
```
Nit (tiny, just for consistency): can you swap `targets()` and `hosts()` around? Usually the ordering is hosts then targets right :D

jieyouxu commented on src/ci/docker/host-x86_64/mingw-check-1/Dockerfile:
```diff
@@ -48,7 +48,7 @@ ENV SCRIPT \
            python3 ../x.py test --stage 0 src/tools/compiletest && \
            python3 ../x.py check compiletest --set build.compiletest-use-stage0-libtest=true && \
-           python3 ../x.py check --stage 1 --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu && \
+           python3 ../x.py check --stage 2 --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu && \
```
Suggestion: actually, can you add a comment for this specifically that we're exercising *cross-compile* here, so `--stage 1` isn't sufficient?

jieyouxu reviewed (commented):
I like this a lot in general, it does feel much more self-consistent and coherent.

I left some questions/discussions for things that I wanted to double-check, but overall looks good!

Kobzol committed:
Make build compiler explicit in `check::Rustc` and `check::Std`

Kobzol committed:
Fix CI

Kobzol committed:
Add staging comment to `check::Rustc`

Kobzol commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -265,8 +276,10 @@ impl Step for CodegenBackend {
         for &backend in &["cranelift", "gcc"] {
-            run.builder.ensure(CodegenBackend { target: run.target, backend });
+            run.builder.ensure(CodegenBackend { build_compiler, target: run.target, backend });
         }
```
Yes, because of the prefix matching logic, `x check compiler` also matches the codegen backends (even if they are not enabled :/). But this doesn't happen for build, because that special cases `compiler` for the `Assemble` step.. a mess, as is usual.

Kobzol commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -371,12 +372,16 @@ impl Step for RustAnalyzer {
+
+    fn metadata(&self) -> Option<StepMetadata> {
+        Some(StepMetadata::check("rust-analyzer", self.target).built_by(self.build_compiler))
+    }
```
:shrug:

jieyouxu replied:
Answer: I don't think so, or at least it's not *as* important as checking the main r-a itself.

Kobzol commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -473,16 +475,33 @@ macro_rules! tool_check_step {
+                    Mode::ToolBootstrap => run.builder.compiler(0, host),
+                    Mode::ToolStd => {
+                        // A small number of tools rely on in-tree standard
+                        // library crates (e.g. compiletest needs libtest).
```
Yes, good point, I'll remove/change the ocmment.

Kobzol commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -515,33 +532,56 @@ fn run_tool_check_step(
         &[],
     );
 
+    // FIXME: check bootstrap doesn't currently work with --all-targets
     cargo.arg("--all-targets");
```
Created it: https://github.com/rust-lang/rust/issues/143385.

Kobzol commented on src/bootstrap/src/core/build_steps/check.rs:
```diff
@@ -253,7 +253,7 @@ fn prepare_compiler_for_check(
             let build_compiler = builder.compiler(builder.top_stage, host);
             builder.std(build_compiler, host);
-            builder.std(build_compiler, target);
+            builder.ensure(Std::new(build_compiler, target));
```
Equivalent to what? :) I changed this just to this:
```rust
let build_compiler = builder.compiler(builder.top_stage, host);
            builder.ensure(Std::new(build_compiler, target));
            build_compiler
```
because I remembered that stdlib's proc macros and build scripts are always compiled using the stage0 libstd. We only need the host stdlib when compiling host code of other modes (yeah, I know..).

jieyouxu replied:
> Equivalent to what?

NGL, I don't know what I meant by this myself -- when I click at this diff, it's outdated against *something*, and I can't tell what it was...

Kobzol committed:
Update `Mode::ToolStd` comment

Kobzol committed:
Update library cross-compilation test

Kobzol committed:
Add `#[tracing::instrument]` to the LLVM build step

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16086115898/job/45397441454) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45397441454) [(plain)](https://github.com/rust-lang/rust/commit/996cd081727d7090fb245673da25e217a276de85/checks/45397441454/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#11 DONE 2.4s

#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 typescript@5.7.3 -g
#12 5.664 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.711 npm WARN deprecated @types/minimatch@6.0.0: This is a stub types definition. minimatch provides its own type definitions, so you do not need this installed.
#12 5.736 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 5.759 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 5.811 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 5.850 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 5.854 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
---
#16 3.065 Building wheels for collected packages: reuse
#16 3.066   Building wheel for reuse (pyproject.toml): started
#16 3.287   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.288   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#16 3.288   Stored in directory: /tmp/pip-ephem-wheel-cache-udcbu0ru/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.290 Successfully built reuse
#16 3.291 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.706 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.707 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.8s
```

</details>

Kobzol committed:
Use stage auto-bump when cross-checking on stage 1

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16086519603/job/45398357258) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45398357258) [(plain)](https://github.com/rust-lang/rust/commit/e38a28d4bac9251042eadf15b50b14cfb16f2dfc/checks/45398357258/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   |
   = note: candidate #1: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_ast-dcd0f45ddadb4315.rmeta
   = note: candidate #2: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_ast-f00d338ea1166c57.rmeta

error[E0464]: multiple candidates for `rmeta` dependency `rustc_codegen_ssa` found
  --> compiler/rustc_codegen_cranelift/src/lib.rs:22:1
   |
22 | extern crate rustc_codegen_ssa;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidate #1: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_codegen_ssa-0a822ae5c81662a5.rmeta
---
   |
   = note: candidate #1: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_index-09b176d543bac198.rmeta
   = note: candidate #2: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_index-bdd4221d2a9f6807.rmeta

error[E0464]: multiple candidates for `rmeta` dependency `rustc_metadata` found
  --> compiler/rustc_codegen_cranelift/src/lib.rs:29:1
   |
29 | extern crate rustc_metadata;
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: candidate #1: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-sysroot/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_metadata-404f084d1be2dbff.rmeta
---

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/abi/mod.rs:270:26
    |
270 |                     _ => bug!("spread argument isn't a tuple?! but {:?}", arg_ty),
    |                          ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/abi/mod.rs:493:18
    |
493 |             _ => bug!("argument to function with \"rust-call\" ABI is not a tuple"),
    |                  ^^^

error: cannot find macro `span_bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:504:57
    |
504 |             TerminatorKind::TailCall { fn_span, .. } => span_bug!(
    |                                                         ^^^^^^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:560:17
    |
560 |                 bug!("shouldn't exist at codegen {:?}", bb_data.terminator());
    |                 ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:668:34
    |
668 | ...                   _ => bug!("Unexpected `PtrToMetadata` operand: {operand:?}"),
    |                            ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:694:30
    |
694 |                         _ => bug!("Trying to ReifyFnPointer on non FnDef {:?}", from_ty),
    |                              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:713:21
    |
713 |                     bug!(
    |                     ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:782:30
    |
782 |                         _ => bug!("{} cannot be cast to a fn ptr", operand.layout().ty),
    |                              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:896:25
    |
896 |                         bug!("RawPtr fields: {operands:?}");
    |                         ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/base.rs:990:14
    |
990 |         _ => bug!("Rvalue::Len({:?})", place),
    |              ^^^

error: cannot find macro `bug` in this scope
    --> compiler/rustc_codegen_cranelift/src/base.rs:1005:42
     |
1005 |             PlaceElem::OpaqueCast(ty) => bug!("encountered OpaqueCast({ty}) in codegen"),
     |                                          ^^^

error: cannot find macro `bug` in this scope
  --> compiler/rustc_codegen_cranelift/src/common.rs:22:17
   |
22 |         bits => bug!("ptr_sized_integer: unknown pointer bit size {}", bits),
   |                 ^^^

error: cannot find macro `bug` in this scope
  --> compiler/rustc_codegen_cranelift/src/common.rs:80:25
   |
80 |         ty::Param(_) => bug!("ty param {:?}", ty),
   |                         ^^^

error: cannot find macro `span_bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/common.rs:466:21
    |
466 |                     span_bug!(span, "`fn_abi_of_fn_ptr({sig}, {extra_args:?})` failed: {err:?}");
    |                     ^^^^^^^^

error: cannot find macro `span_bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/common.rs:469:21
    |
469 |                     span_bug!(
    |                     ^^^^^^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/debuginfo/mod.rs:268:74
    |
268 |         let DefKind::Static { nested, .. } = tcx.def_kind(def_id) else { bug!() };
    |                                                                          ^^^

error: cannot find macro `span_bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/inline_asm.rs:117:21
    |
117 |                     span_bug!(span, "invalid type for asm sym (fn)");
    |                     ^^^^^^^^

error: cannot find macro `span_bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/inline_asm.rs:126:17
    |
126 |                 span_bug!(span, "asm! label operands are not yet supported");
    |                 ^^^^^^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/llvm_x86.rs:201:22
    |
201 |                 _ => bug!("wrong number of args for intrinsic {intrinsic}"),
    |                      ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/llvm_x86.rs:314:22
    |
314 |                 _ => bug!("wrong number of args for intrinsic {intrinsic}"),
    |                      ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/llvm_x86.rs:367:22
    |
367 |                 _ => bug!("wrong number of args for intrinsic {intrinsic}"),
    |                      ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs:122:17
    |
122 |                 bug!("wrong number of args for intrinsic {intrinsic}");
    |                 ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs:167:21
    |
167 |                     bug!("wrong number of args for intrinsic {intrinsic}");
    |                     ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs:260:21
    |
260 |                     bug!("wrong number of args for intrinsic {intrinsic}");
    |                     ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/simd.rs:305:21
    |
305 |                     bug!("wrong number of args for intrinsic {intrinsic}");
    |                     ^^^

error: cannot find macro `bug` in this scope
  --> compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs:34:5
   |
34 |     bug!("wrong number of args for intrinsic {}", intrinsic);
   |     ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs:395:9
    |
395 |         bug!("wrong number of args for intrinsic {:?}", intrinsic);
    |         ^^^

error: cannot find macro `span_bug` in this scope
    --> compiler/rustc_codegen_cranelift/src/intrinsics/mod.rs:1469:17
     |
1469 |                 span_bug!(
     |                 ^^^^^^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/num.rs:300:14
    |
300 |         _ => bug!("binop {:?} on checked int/uint lhs: {:?} rhs: {:?}", bin_op, in_lhs, in_rhs),
    |              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/num.rs:389:22
    |
389 |                 _ => bug!(),
    |                      ^^^

error: cannot find macro `bug` in this scope
  --> compiler/rustc_codegen_cranelift/src/unsize.rs:70:14
   |
70 |         _ => bug!("unsized_info: invalid unsizing {:?} -> {:?}", source, target),
   |              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/unsize.rs:111:14
    |
111 |         _ => bug!("unsize_ptr: called on bad types"),
    |              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/unsize.rs:158:14
    |
158 |         _ => bug!("coerce_unsized_into: invalid coercion {:?} -> {:?}", src_ty, dst_ty),
    |              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/unsize.rs:289:14
    |
289 |         _ => bug!("size_and_align_of_dst: {ty} not supported"),
    |              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:151:47
    |
151 |             CValueInner::ByRef(_, Some(_)) => bug!("load_scalar for unsized value not allowed"),
    |                                               ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:152:45
    |
152 |             CValueInner::ByValPair(_, _) => bug!("Please use load_scalar_pair for ByValPair"),
    |                                             ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:176:17
    |
176 |                 bug!("load_scalar_pair for unsized value not allowed")
    |                 ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:178:38
    |
178 |             CValueInner::ByVal(_) => bug!("Please use load_scalar for ByVal"),
    |                                      ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:196:30
    |
196 |                         _ => bug!("field should be 0 or 1"),
    |                              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:483:46
    |
483 |             CPlaceInner::Addr(_, Some(_)) => bug!("Expected sized cplace, found {:?}", self),
    |                                              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:485:17
    |
---

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:504:46
    |
504 |             CPlaceInner::Addr(_, Some(_)) => bug!("Expected sized cplace, found {:?}", self),
    |                                              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:593:46
    |
593 |             CPlaceInner::Addr(_, Some(_)) => bug!("Can't write value to unsized place {:?}", self),
    |                                              ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:614:29
    |
---

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:802:56
    |
802 |                     | CPlaceInner::VarPair(_, _, _) => bug!("Can't index into {self:?}"),
    |                                                        ^^^

error: cannot find macro `bug` in this scope
   --> compiler/rustc_codegen_cranelift/src/value_and_place.rs:806:18
    |
806 |             _ => bug!("place_index({:?})", self.layout().ty),
    |                  ^^^

error: cannot find macro `info` in this scope
   --> compiler/rustc_codegen_cranelift/src/lib.rs:241:9
    |
241 |         info!("codegen crate {}", tcx.crate_name(LOCAL_CRATE));
    |         ^^^^

Some errors have detailed explanations: E0432, E0464.
For more information about an error, try `rustc --explain E0432`.
[RUSTC-TIMING] rustc_codegen_cranelift test:false 10.525
```

</details>

Kobzol committed:
Add support for allowing features when checking tools

Kobzol committed:
Horrible hack to make codegen backends "work" during check

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16087328986/job/45400084412) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45400084412) [(plain)](https://github.com/rust-lang/rust/commit/dddb831ebb3efce799750cf6e84135ce51dd96ca/checks/45400084412/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] build_helper test:false 0.669
error: the borrowed expression implements the required traits
   --> src/bootstrap/src/core/build_steps/check.rs:343:40
    |
343 |         let _guard = builder.msg_check(&format!("rustc_codegen_{backend}"), target, None);
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `format!("rustc_codegen_{backend}")`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrows_for_generic_args
    = note: `-D clippy::needless-borrows-for-generic-args` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::needless_borrows_for_generic_args)]`

```

</details>

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16091659227/job/45409335178) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45409335178) [(plain)](https://github.com/rust-lang/rust/commit/263f2024532550a15894161dc5c8b9857777ee40/checks/45409335178/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#11 DONE 2.5s

#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 typescript@5.7.3 -g
#12 6.943 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 6.967 npm WARN deprecated @types/minimatch@6.0.0: This is a stub types definition. minimatch provides its own type definitions, so you do not need this installed.
#12 6.982 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 7.002 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 7.042 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 7.079 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
#12 7.091 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
---
#16 3.492 Building wheels for collected packages: reuse
#16 3.493   Building wheel for reuse (pyproject.toml): started
#16 3.710   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.711   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#16 3.712   Stored in directory: /tmp/pip-ephem-wheel-cache-qwo0bnex/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.714 Successfully built reuse
#16 3.715 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 4.110 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 4.111 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 4.2s
```

</details>

rust-log-analyzer commented:
The job **`pr-check-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16092130703/job/45410311267) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45410311267) [(plain)](https://github.com/rust-lang/rust/commit/49cc81d352f6c365f13ce65a920152f4b9f83d7f/checks/45410311267/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#11 DONE 2.3s

#12 [ 5/11] RUN npm install es-check@6.1.1 eslint@8.6.0 typescript@5.7.3 -g
#12 6.108 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 6.140 npm WARN deprecated @types/minimatch@6.0.0: This is a stub types definition. minimatch provides its own type definitions, so you do not need this installed.
#12 6.157 npm WARN deprecated rimraf@3.0.2: Rimraf versions prior to v4 are no longer supported
#12 6.166 npm WARN deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
#12 6.216 npm WARN deprecated @humanwhocodes/config-array@0.9.5: Use @eslint/config-array instead
#12 6.251 npm WARN deprecated @humanwhocodes/object-schema@1.2.1: Use @eslint/object-schema instead
#12 6.271 npm WARN deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
---
#16 2.940 Building wheels for collected packages: reuse
#16 2.941   Building wheel for reuse (pyproject.toml): started
#16 3.154   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.155   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=d2a2565e7037ad3883fb9337653f2e25bbb588534fbef3697286cbc26d1bf634
#16 3.155   Stored in directory: /tmp/pip-ephem-wheel-cache-8lhndd34/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.158 Successfully built reuse
#16 3.158 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.555 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.555 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
```

</details>

Kobzol committed:
Do not ever cross-check bootstrap tools

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16092477413/job/45411044570) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/45411044570) [(plain)](https://github.com/rust-lang/rust/commit/cf0eddb2f1def7ee6e1c8ee2611f53cec39bba29/checks/45411044570/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[group]Checking tool miri (stage1:aarch64-unknown-linux-gnu -> stage2:i686-unknown-haiku)
##[endgroup]
##[group]Checking tool cargo-miri (stage1:aarch64-unknown-linux-gnu -> stage2:i686-unknown-haiku)
##[endgroup]
WARNING: not checking bootstrap tool "src/tools/miropt-test-tools" for target i686-unknown-haiku as it is a bootstrap (host-only) tool
##[group]Checking tool rustfmt (stage1:aarch64-unknown-linux-gnu -> stage2:i686-unknown-haiku)
##[endgroup]
##[group]Checking stage1 rust-analyzer artifacts (aarch64-unknown-linux-gnu -> i686-unknown-haiku)
##[endgroup]
##[group]Checking stage2 tool test-float-parse (aarch64-unknown-linux-gnu -> i686-unknown-haiku)
##[endgroup]
WARNING: not checking bootstrap tool "src/tools/features-status-dump" for target i686-unknown-haiku as it is a bootstrap (host-only) tool
##[group]Checking stage1 library artifacts (aarch64-unknown-linux-gnu -> i686-unknown-haiku)
##[endgroup]
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ Snapshot Summary ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Snapshot: check_cross_compile
Source: src/bootstrap/src/core/builder/tests.rs:1300
────────────────────────────────────────────────────────────────────────────────
Expression: ctx.config("check").targets(&[TEST_TRIPLE_1]).hosts(&[TEST_TRIPLE_1]).render_steps()
────────────────────────────────────────────────────────────────────────────────
-old snapshot
+new results
────────────┬───────────────────────────────────────────────────────────────────
    7     7 │ [check] rustc 1 <host> -> gcc 2 <target1>
    8     8 │ [check] rustc 1 <host> -> Clippy 2 <target1>
    9     9 │ [check] rustc 1 <host> -> Miri 2 <target1>
   10    10 │ [check] rustc 1 <host> -> CargoMiri 2 <target1>
   11       │-[check] rustc 0 <host> -> MiroptTestTools 1 <target1>
   12    11 │ [check] rustc 1 <host> -> Rustfmt 2 <target1>
   13    12 │ [check] rustc 1 <host> -> rust-analyzer 2 <target1>
   14    13 │ [check] rustc 1 <host> -> TestFloatParse 2 <target1>
   15       │-[check] rustc 0 <host> -> FeaturesStatusDump 1 <target1>
   16    14 │ [check] rustc 1 <host> -> std 1 <target1>
────────────┴───────────────────────────────────────────────────────────────────
Stopped on the first failure. Run `cargo insta test` to run all snapshots.

thread 'core::builder::tests::snapshot::check_cross_compile' panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/insta-1.43.1/src/runtime.rs:679:13:
snapshot assertion for 'check_cross_compile' failed in line 1300
```

</details>

jieyouxu commented:
@rustbot review

jieyouxu commented on src/bootstrap/src/utils/change_tracker.rs:
```diff
@@ -446,4 +446,9 @@ pub const CONFIG_CHANGE_HISTORY: &[ChangeInfo] = &[
+    ChangeInfo {
+        change_id: 143048,
+        severity: ChangeSeverity::Warning,
+        summary: "The default check stage has been changed to 1. It is no longer possible to `x check` with stage 0. All check commands have to be on stage 1+.",
```
Discussion: maybe also mention cross-compile check of (rustc) must also be at stage 2?

Kobzol replied:
> Cross-compile check of (rustc) must also be at stage 2

Well, that's not strictly true. We now auto-bump and when you ask for stage 1 cross-compile check, bootstrap will silently do a stage 2 check.

jieyouxu replied:
Ugh, right.

jieyouxu reviewed (approved):
Thanks. The hacks are... quite hacky, but I don't have an alternative scheme that's better without their own problems, so I'm inclined to land this set of consistency fixes first.

Feel free to r=me after the change tracker nit.

bors commented:
:umbrella: The latest upstream changes (presumably #143521) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Kobzol commented:
My rebases lost author information of Zalathar from the code I reused from https://github.com/rust-lang/rust/pull/139170, reinstated that. Rebased and updated the change tracker.

Kobzol commented:
@bors r=jieyouxu rollup=never

bors commented:
:hourglass: Testing commit 5dc77ba6586ded662d4c5447da87a70d8a77651a with merge c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2...
<!-- homu: {"type":"BuildStarted","head_sha":"5dc77ba6586ded662d4c5447da87a70d8a77651a","merge_sha":"c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust/actions/runs/16108066983/job/45455979569)
Approved by: jieyouxu
Pushing c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jieyouxu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust/actions/runs/16108066983/job/45455979569"},"merge_sha":"c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2"} -->

rust-timer commented:
Finished benchmarking commit (c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2): [comparison URL](https://perf.rust-lang.org/compare.html?start=0d11be5aabe0cd49609fff5fce57c4691a22fe55&end=c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=0d11be5aabe0cd49609fff5fce57c4691a22fe55&end=c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.4%, -1.4%] | 1     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-1.4%, -0.3%] | 2     |
| All ❌✅ (primary)                 | -1.4% | [-1.4%, -1.4%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=0d11be5aabe0cd49609fff5fce57c4691a22fe55&end=c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2&stat=max-rss)
<details>
<summary>Results (secondary 2.0%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.0% | [1.7%, 2.4%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=0d11be5aabe0cd49609fff5fce57c4691a22fe55&end=c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2&stat=cycles:u)
<details>
<summary>Results (secondary 2.4%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.4% | [2.4%, 2.4%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=0d11be5aabe0cd49609fff5fce57c4691a22fe55&end=c720f49c46ec3e70bbc2bbb3e0dd1fe7fc0a47f2&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 462.394s -> 464.259s (0.40%)
**Artifact size**: 372.15 MiB -> 372.10 MiB (-0.02%)

matthiaskrgr commented:
How can I do just a `check` build now?

I had 
```toml

# The default stage to use for the `check` subcommand
check-stage = 0

# The default stage to use for the `doc` subcommand
doc-stage = 0

# The default stage to use for the `build` subcommand
build-stage = 1

# The default stage to use for the `test` subcommand
test-stage = 1

# The default stage to use for the `dist` subcommand
dist-stage = 2

# The default stage to use for the `install` subcommand
install-stage = 2

# The default stage to use for the `bench` subcommand
bench-stage = 2
````

which no longer works now:

```
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
WARNING: cannot check anything on stage 0. Use at least stage 1.
Build completed unsuccessfully in 0:00:00
```

so I set
```toml
# The default stage to use for the `check` subcommand
check-stage = 1

# The default stage to use for the `doc` subcommand
doc-stage = 1
...
```

which now does  a check build but also a full `stage0 -> stage1` bootstrap right after?? :thinking: 

```
./x.py check
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
Checking stage1 compiler artifacts {rustc-main, rustc_abi, rustc_arena, rustc_ast, rustc_ast_ir, rustc_ast_lowering, rustc_ast_passes, rustc_ast_pretty, rustc_attr_data_structures, rustc_attr_parsing, rustc_baked_icu_data, rustc_borrowck, rustc_builtin_macros, rustc_codegen_llvm, rustc_codegen_ssa, rustc_const_eval, rustc_data_structures, rustc_driver, rustc_driver_impl, rustc_error_codes, rustc_error_messages, rustc_errors, rustc_expand, rustc_feature, rustc_fluent_macro, rustc_fs_util, rustc_graphviz, rustc_hashes, rustc_hir, rustc_hir_analysis, rustc_hir_pretty, rustc_hir_typeck, rustc_incremental, rustc_index, rustc_index_macros, rustc_infer, rustc_interface, rustc_lexer, rustc_lint, rustc_lint_defs, rustc_llvm, rustc_log, rustc_macros, rustc_metadata, rustc_middle, rustc_mir_build, rustc_mir_dataflow, rustc_mir_transform, rustc_monomorphize, rustc_next_trait_solver, rustc_parse, rustc_parse_format, rustc_passes, rustc_pattern_analysis, rustc_privacy, rustc_proc_macro, rustc_query_impl, rustc_query_system, rustc_resolve, rustc_sanitizers, rustc_serialize, rustc_session, rustc_smir, rustc_span, rustc_symbol_mangling, rustc_target, rustc_thread_pool, rustc_trait_selection, rustc_traits, rustc_transmute, rustc_ty_utils, rustc_type_ir, rustc_type_ir_macros, stable_mir} (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.29s
Checking tool rustdoc (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.15s
Checking stage1 rustc_codegen_cranelift (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.05s
Checking stage1 rustc_codegen_gcc (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.03s
Checking tool clippy (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.20s
Checking tool miri (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.18s
Checking tool cargo-miri (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.14s
Checking stage1 tool miropt-test-tools (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.11s
Checking tool rustfmt (stage0 -> stage1, x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.17s
Checking stage1 rust-analyzer artifacts (x86_64-unknown-linux-gnu)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.30s
/home/matthias/vcs/github/rust/build/x86_64-unknown-linux-gnu/ci-llvm/bin/llvm-strip does not exist; skipping copy
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
   Compiling rustc_llvm v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_llvm)
   Compiling rustc_parse v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_parse)
   Compiling rustc_middle v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_middle)
   Compiling rustc_ast_passes v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_ast_passes)
   Compiling rustc_expand v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_expand)
   Compiling rustc_builtin_macros v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_builtin_macros)
   Compiling rustc_infer v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_infer)
   Compiling rustc_transmute v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_transmute)
   Compiling rustc_incremental v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_incremental)
   Compiling rustc_symbol_mangling v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_symbol_mangling)
   Compiling rustc_mir_dataflow v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_mir_dataflow)
   Compiling rustc_pattern_analysis v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_pattern_analysis)
   Compiling rustc_ast_lowering v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_ast_lowering)
   Compiling rustc_query_impl v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_query_impl)
   Compiling rustc_smir v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_smir)
   Compiling stable_mir v0.1.0-preview (/home/matthias/vcs/github/rust/compiler/stable_mir)
   Compiling rustc_monomorphize v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_monomorphize)
   Compiling rustc_metadata v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_metadata)
   Compiling rustc_trait_selection v0.0.0 (/home/matthias/vcs/github/rust/compiler/rustc_trait_selection)
```

Kobzol commented:
`x check` checks both rustc and std, amongst other things. For checking std, you need to *build* a stage 1 rustc after the stage0 redesign, there's no way around that (other than download-ci-rustc).
